app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt

```diff
@@ -229,78 +146,28 @@ class TabActivity : BaseActivity() {
                 setAction(R.string.reading_list_item_delete_undo) {
-                    app.tabList.add(appTabIndex, appTab)
-                    binding.tabSwitcher.addTab(tab, index)
+                    app.tabList.add(index, appTab)
+                    binding.tabRecyclerView.adapter?.notifyItemInserted(adapterPosition)
+                    binding.tabCountsView.updateTabCount(false)
                 }
                 show()
             }
         }
     }
 
-    private fun showUndoAllSnackbar(tabs: Array<de.mrapp.android.tabswitcher.Tab>, appTabs: MutableList<Tab>) {
+    private fun showUndoAllSnackbar(appTabs: MutableList<Tab>) {
         FeedbackUtil.makeSnackbar(this, getString(R.string.all_tab_items_closed)).run {
             setAction(R.string.reading_list_item_delete_undo) {
                 app.tabList.addAll(appTabs)
-                binding.tabSwitcher.addAllTabs(tabs)
                 appTabs.clear()
-            }
-            show()
-        }
-    }
-
-    private inner class TabListener : TabSwitcherListener {
-        override fun onSwitcherShown(tabSwitcher: TabSwitcher) {}
-        override fun onSwitcherHidden(tabSwitcher: TabSwitcher) {}
-        override fun onSelectionChanged(tabSwitcher: TabSwitcher, index: Int, selectedTab: de.mrapp.android.tabswitcher.Tab?) {
-            if (app.tabList.isNotEmpty() && index < app.tabList.size) {
-                val tabIndex = app.tabList.size - index - 1
-                L.d("Tab selected: $index")
-                if (tabIndex < app.tabList.size - 1) {
-                    val tab = app.tabList.removeAt(tabIndex)
-                    app.tabList.add(tab)
-                }
-                binding.tabCountsView.updateTabCount(false)
-                cancelled = false
-                val tabUpdateDebounceMillis = 250
-                if (System.currentTimeMillis() - tabUpdatedTimeMillis > tabUpdateDebounceMillis) {
-                    if (launchedFromPageActivity) {
-                        setResult(RESULT_LOAD_FROM_BACKSTACK)
-                    } else {
-                        startActivity(PageActivity.newIntent(this@TabActivity))
-                    }
-                    finish()
-                }
-            }
-        }
-
-        override fun onTabAdded(tabSwitcher: TabSwitcher, index: Int, tab: de.mrapp.android.tabswitcher.Tab, animation: Animation) {
-            binding.tabCountsView.updateTabCount(false)
-            tabUpdatedTimeMillis = System.currentTimeMillis()
-        }
-
-        override fun onTabRemoved(tabSwitcher: TabSwitcher, index: Int, tab: de.mrapp.android.tabswitcher.Tab, animation: Animation) {
-            if (app.tabList.isNotEmpty() && index < app.tabList.size) {
-                val tabIndex = app.tabList.size - index - 1
-                val appTab = app.tabList.removeAt(tabIndex)
+                binding.tabRecyclerView.adapter?.notifyDataSetChanged()
```

Comment:
Would it be a good idea to use `binding.tabRecyclerView.adapter?.notifyItemRangeChanged(0, app.tabList.size)` for a smoother animation?

Reply:
yep